### PR TITLE
Battle: color cycling for glowing effects

### LINKED ISSUE #505: A lightning in Titan's/Giant's hand
In the OG a lightning in Titan's or Giant's hand is glowing.
Titan

### LINKED ISSUE #508: Phoenix's glowing
In the original game Phoenix's feathers glow while idle.

### LINKED ISSUE #535: Creatures don't have animated or glowing effects
Some creatures don't have animated or glowing effects, that they have in original game.
This issue relates to:
#508 
#505 
#536 
#542
#593

### LINKED ISSUE #536: Genies don't have animated affect over their bottom
In the original game Genies have spinning effect of their bottom.
Genie

### LINKED ISSUE #542: Fire Elemental doesn't have glowing effect
In the original game Fire Elemetal glows while idle.

### LINKED ISSUE #593: Top of mage's staff doesn't have glowing effect
There is tiny glowing crystal on he top of Mage and Archmage's staff in the OG.
Our version is missing this efect.

### PR DESCRIPTION
**Pull request by idshibanov**
Fixes #535. Fixes #505. Fixes #508 . Fixes #536. Fixes #542. Fixes #593.

Adventure map effects will be in a separate PR.

ezgif-5-61a03558a005

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @idshibanov , everything looks cool but now it enormously slow in comparison before changes. Give me some time to think about a way how to speed things up...

**Comment by ihhub** [comment]
Hi @idshibanov , could you please fix compilation on Unix systems and test this function changes:
```
Surface Surface::GetSurface( void ) const
{
    return Surface( *this, false );
}
```
It is not a good optimization but I think it is slightly faster than current implementation. I will continue looking for places where we can optimize it.

**Comment by ihhub** [comment]
Another enhancement: in `Surface::RenderChangeColor` please add this code for 32-bit images just before the loop:
```
        if ( pitch == width ) {
            width = width * height;
            pitch = width;
            height = 1;
        }
```
then we will make less computations in first loop. Not a big deal but a small improvement.

**Comment by ihhub** [comment]
The main reason of slow drawings are in STL extra checks for Debug builds in Windows. I will make needed changes in this pull request to avoid them.

**Comment by ihhub** [comment]
Hi @idshibanov , I made all needed optimization. The only thing left is to add a special function which would return **true** only for monsters with animation so we do not waste our CPU resources for unneeded false drawings. Please add this function.

File: src/fheroes2/battle/battle_interface.cpp [hunk 1]
```diff
@@ -1235,6 +1253,8 @@ void Battle::Interface::RedrawTroopSprite( const Unit & b ) const
             sp.y += cy + ( _movingPos.y - _flyingPos.y ) * _flyingUnit->animation.movementProgress();
         }
 
+        spmon1.ChangeColor( _colorCyclePairs );
```
Comment (ihhub) [inline src/fheroes2/battle/battle_interface.cpp:1256]: Here we need to check if the monster has a specific animation then draw if any.

Comment (ihhub) [reply src/fheroes2/battle/battle_interface.cpp:1256]: Also we should not apply same color change for troops with `SP_STONE` effect.

**Comment by idshibanov** [comment]
> Hi @idshibanov , I made all needed optimization. The only thing left is to add a special function which would return **true** only for monsters with animation so we do not waste our CPU resources for unneeded false drawings. Please add this function.

Perfect! Thanks, I knew you would add those optimizations :) I'll fix the rest

**Comment by ihhub** [comment]
Hi @idshibanov , excellent work! Thanks a lot!

**Pull request merged**